async-usercalls/src/io_bufs.rs

```diff
@@ -0,0 +1,259 @@
+    /// Panics if called more than once in a row without either calling `consume()`
+    /// or dropping the previously returned buffer.
+    pub fn consumable_chunk(&mut self) -> Option<UserBuf> {
+        assert!(
+            Arc::strong_count(&self.userbuf) == 1,
+            "called consumable_chunk() more than once in a row"
+        );
+        let range = match self.offsets() {
+            (_, _) if self.read == self.write => return None, // empty
+            (r, w) if r < w => r..w,
+            (r, _) => r..self.buf_len,
+        };
+        Some(UserBuf(UserBufKind::Shared {
+            user: self.userbuf.clone(),
+            range,
+        }))
+    }
+
+    /// Mark `n` bytes as consumed. `buf` must have been produced by a call
+    /// to `self.consumable_chunk()`.
+    /// Panics if:
+    /// - `n > buf.len()`
+    /// - `buf` was not produced by `self.consumable_chunk()`
+    ///
+    /// This function is supposed to be used in conjunction with `consumable_chunk()`.
+    pub fn consume(&mut self, buf: UserBuf, n: usize) {
+        assert!(n <= buf.len());
+        const PANIC_MESSAGE: &'static str = "`buf` not produced by self.consumable_chunk()";
+        let buf = buf.into_shared().expect(PANIC_MESSAGE);
+        assert!(Arc::ptr_eq(&self.userbuf, &buf), PANIC_MESSAGE);
+        drop(buf);
+        assert!(Arc::strong_count(&self.userbuf) == 1, PANIC_MESSAGE);
+        self.advance_read(n);
+    }
+
+    fn len(&self) -> usize {
+        match self.offsets() {
+            (_, _) if self.read == self.write => 0,                      // empty
+            (r, w) if r == w && self.read != self.write => self.buf_len, // full
+            (r, w) if r < w => w - r,
+            (r, w) => w + self.buf_len - r,
+        }
+    }
+
+    fn remaining_capacity(&self) -> usize {
+        let len = self.len();
+        debug_assert!(len <= self.buf_len);
+        self.buf_len - len
+    }
+
+    fn offsets(&self) -> (usize, usize) {
+        (self.read as usize % self.buf_len, self.write as usize % self.buf_len)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.read == self.write
+    }
+
+    fn is_full(&self) -> bool {
+        let (read_offset, write_offset) = self.offsets();
+        read_offset == write_offset && self.read != self.write
+    }
+
+    fn advance_read(&mut self, by: usize) {
+        debug_assert!(by <= self.len());
+        self.read = ((self.read as usize + by) % (self.buf_len * 2)) as _;
+    }
+
+    fn advance_write(&mut self, by: usize) {
+        debug_assert!(by <= self.remaining_capacity());
+        self.write = ((self.write as usize + by) % (self.buf_len * 2)) as _;
+    }
+}
+
+pub struct ReadBuffer {
+    userbuf: User<[u8]>,
+    position: usize,
+    len: usize,
+}
+
+impl ReadBuffer {
+    /// Constructs a new `ReadBuffer`, assuming `len` bytes of `userbuf` have
+    /// meaningful data. Panics if `len > userbuf.len()`.
+    pub fn new(userbuf: User<[u8]>, len: usize) -> ReadBuffer {
```

Comment:
Is this `len` parameter required? Can't it simply use the lenght of the provided buffer? `WriteBuffer` doesn't have such a parameter.

Reply:
I think the comments on top of the function explain why the parameter is there. Here is the reason behind it: to read from a socket, one needs to allocate a userspace buffer first and then call read() or send a read usercall asynchronously. In either case, the read usercall returns a number indicating how many bytes were actually read. To deliver these bytes to code running inside the enclave, one would construct an instance of `ReadBuffer` passing in the userspace buffer and the number of bytes returned from read.